### PR TITLE
service/elbv2/load_balancer: Support WAF fail open

### PR DESCRIPTION
**Pull request by shuheiktgw**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes https://github.com/hashicorp/terraform-provider-aws/issues/16372

Release note for [CHANGELOG](https://github.com/hashicorp/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
resoure/aws_lb: Support WAF fial open
```

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccAWSLB_applicationLoadBalancer_updateWafFailOpen'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -count 1 -parallel 20 -run=TestAccAWSLB_applicationLoadBalancer_updateWafFailOpen -timeout 120m
=== RUN   TestAccAWSLB_applicationLoadBalancer_updateWafFailOpen
=== PAUSE TestAccAWSLB_applicationLoadBalancer_updateWafFailOpen
=== CONT  TestAccAWSLB_applicationLoadBalancer_updateWafFailOpen
--- PASS: TestAccAWSLB_applicationLoadBalancer_updateWafFailOpen (639.39s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	641.383s

```

Support WAF fail open for ALBs. Thank you for your review! 👍

### CODE REVIEW

File: aws/data_source_aws_lb_test.go [hunk 1]
```diff
@@ -117,6 +117,7 @@ func TestAccDataSourceAWSLB_BackwardsCompatibility(t *testing.T) {
 					resource.TestCheckResourceAttrPair(dataSourceName1, "subnet_mapping.#", resourceName, "subnet_mapping.#"),
 					resource.TestCheckResourceAttrPair(dataSourceName1, "drop_invalid_header_fields", resourceName, "drop_invalid_header_fields"),
 					resource.TestCheckResourceAttrPair(dataSourceName1, "enable_http2", resourceName, "enable_http2"),
+					resource.TestCheckResourceAttrPair(dataSourceName1, "enable_waf_fail_open", resourceName, "enable_http2"),
```
Comment (nunofernandes) [inline aws/data_source_aws_lb_test.go:120]: The resource name doesn't seem right

**Review by nunofernandes — COMMENTED** [review]
Just some copy paste error...

File: aws/data_source_aws_lb_test.go (hunk 1 above)
Comment (shuheiktgw) [reply aws/data_source_aws_lb_test.go:120]: Whoops, thank you!

**Comment by nunofernandes** [comment]
Any news on this PR? Is there anything that needs to be addressed for this to be merged?

**Comment by breathingdust** [comment]
Hi all 👋  Just letting you know that this is issue is featured on this quarters roadmap. If a PR exists to close the issue a maintainer will review and either make changes directly, or work with the original author to get the contribution merged. If you have written a PR to resolve the issue please ensure the "Allow edits from maintainers" box is checked. Thanks for your patience and we are looking forward to getting this merged soon!

**Comment by anGie44** [comment]
Thanks so much for this PR @shuheiktgw ! I've merged in the `main` branch and made some updates to align with the new service package structure"

Output of acceptance tests:
```
--- SKIP: TestAccELBV2LoadBalancer_ALB_outpost (1.43s)
--- PASS: TestAccELBV2LoadBalancer_LoadBalancerTypeGateway_enableCrossZoneLoadBalancing (164.59s)
--- PASS: TestAccELBV2LoadBalancer_namePrefix (174.79s)
--- PASS: TestAccELBV2LoadBalancer_ApplicationLoadBalancer_updateWafFailOpen (281.58s)
--- PASS: TestAccELBV2LoadBalancer_noSecurityGroup (173.68s)
--- PASS: TestAccELBV2LoadBalancer_ALB_basic (185.24s)
--- PASS: TestAccELBV2LoadBalancer_ApplicationLoadBalancer_updateDeletionProtection (285.21s)
--- PASS: TestAccELBV2LoadBalancer_generatedName (175.06s)
--- PASS: TestAccELBV2LoadBalancer_networkLoadBalancerEIP (316.98s)
--- PASS: TestAccELBV2LoadBalancer_NLB_privateIPv4Address (671.40s)
--- PASS: TestAccELBV2LoadBalancer_NetworkLoadBalancerSubnet_change (702.42s)
--- PASS: TestAccELBV2LoadBalancer_generatesNameForZeroValue (735.59s)
--- PASS: TestAccELBV2LoadBalancer_backwardsCompatibility (473.36s)
--- PASS: TestAccELBV2LoadBalancer_NetworkLoadBalancer_updateCrossZone (629.43s)
--- PASS: TestAccELBV2LoadBalancer_NLB_basic (275.09s)
--- PASS: TestAccELBV2LoadBalancer_ipv6SubnetMapping (549.97s)
--- PASS: TestAccELBV2LoadBalancer_updatedSubnets (1083.24s)
--- PASS: TestAccELBV2LoadBalancer_updatedSecurityGroups (1122.58s)
--- PASS: TestAccELBV2LoadBalancer_updatedIPAddressType (1176.44s)
--- PASS: TestAccELBV2LoadBalancer_ApplicationLoadBalancer_updateHTTP2 (1220.89s)
--- PASS: TestAccELBV2LoadBalancer_ApplicationLoadBalancer_updateDropInvalidHeaderFields (1253.31s)
--- PASS: TestAccELBV2LoadBalancer_ALBAccessLogs_prefix (1314.52s)
--- PASS: TestAccELBV2LoadBalancer_NLBAccessLogs_prefix (1315.63s)
--- PASS: TestAccELBV2LoadBalancer_ALB_accessLogs (1350.83s)
--- PASS: TestAccELBV2LoadBalancer_tags (833.49s)
--- PASS: TestAccELBV2LoadBalancer_NLB_accessLogs (1427.57s)
--- PASS: TestAccELBV2LoadBalancer_LoadBalancerType_gateway (132.21s)

--- SKIP: TestAccELBV2LoadBalancerDataSource_outpost (1.45s)
--- PASS: TestAccELBV2LoadBalancerDataSource_basic (245.26s)
--- PASS: TestAccELBV2LoadBalancerDataSource_backwardsCompatibility (253.76s)
```

**Review by anGie44 — APPROVED** [review]

**Comment by wwwizards** [comment]
beginning in v3.67 and continuing in v3.68 there are intermittent errors when creating an ALB

╷
│ Error: failure configuring LB attributes: ValidationError: Load balancer attribute key 'waf.fail_open.enabled' is not recognized
│       status code: 400, request id: b51c6db6-4da1-4727-859f-8666663dc153
│
│   with aws_alb.alb01,
│   on alb01.tf line 4, in resource "aws_alb" "alb01":
│    4: resource "aws_alb" "alb01" {
│

I say "intermittent" because there have been some successful runs but I cannot discern any pattern. Here is some more info from the TRACE log. 

-----------------------------------------------------: timestamp=2021-12-02T11:07:32.818Z
2021-12-02T11:07:32.818Z [INFO]  provider.terraform-provider-aws_v3.68.0_x5: 2021/12/02 11:07:32 [DEBUG] [aws-sdk-go] <?xml version="1.0" encoding="UTF-8"?>
<DescribeNetworkInterfacesResponse xmlns="http://ec2.amazonaws.com/doc/2016-11-15/">
    <requestId>965736a9-2de3-48b8-8588-c19d61d01834</requestId>
    <networkInterfaceSet/>
</DescribeNetworkInterfacesResponse>: timestamp=2021-12-02T11:07:32.818Z
2021-12-02T11:07:32.818Z [TRACE] NodeAbstractResouceInstance.writeResourceInstanceState to workingState for aws_alb.alb01
2021-12-02T11:07:32.818Z [TRACE] NodeAbstractResouceInstance.writeResourceInstanceState: removing state object for aws_alb.alb01
2021-12-02T11:07:32.819Z [TRACE] vertex "aws_alb.alb01 (destroy)": visit complete
2021-12-02T11:07:32.819Z [TRACE] vertex "aws_alb.alb01": starting visit (*terraform.NodeApplyableResourceInstance)
2021-12-02T11:07:32.819Z [TRACE] readDiff: Read DeleteThenCreate change from plan for aws_alb.alb01
2021-12-02T11:07:32.819Z [TRACE] readResourceInstanceState: reading state for aws_alb.alb01
2021-12-02T11:07:32.819Z [TRACE] readResourceInstanceState: no state present for aws_alb.alb01
2021-12-02T11:07:32.819Z [TRACE] readDiff: Read DeleteThenCreate change from plan for aws_alb.alb01
2021-12-02T11:07:32.821Z [TRACE] Re-validating config for "aws_alb.alb01"
2021-12-02T11:07:32.821Z [TRACE] GRPCProvider: ValidateResourceConfig
2021-12-02T11:07:32.822Z [TRACE] GRPCProvider: PlanResourceChange
2021-12-02T11:07:32.825Z [WARN]  Provider "registry.terraform.io/hashicorp/aws" produced an invalid plan for aws_alb.alb01, but we are tolerating it because it is using the legacy plugin SDK.
    The following problems may be the cause of any confusing errors from downstream operations:
      - .drop_invalid_header_fields: planned value cty.False for a non-computed attribute
      - .idle_timeout: planned value cty.NumberIntVal(60) for a non-computed attribute
      - .enable_deletion_protection: planned value cty.False for a non-computed attribute
      - .load_balancer_type: planned value cty.StringVal("application") for a non-computed attribute
      - .desync_mitigation_mode: planned value cty.StringVal("defensive") for a non-computed attribute
      - .enable_http2: planned value cty.True for a non-computed attribute
      - .enable_waf_fail_open: planned value cty.False for a non-computed attribute
      - .subnet_mapping: attribute representing nested block must not be unknown itself; set nested attribute values to unknown instead
2021-12-02T11:07:32.825Z [TRACE] plan: aws_alb.alb01 treating Create change as DeleteThenCreate change to match with earlier plan
2021-12-02T11:07:32.825Z [TRACE] checkPlannedChange: Verifying that actual change (action DeleteThenCreate) matches planned change (action DeleteThenCreate)
2021-12-02T11:07:32.825Z [TRACE] readResourceInstanceState: reading state for aws_alb.alb01
2021-12-02T11:07:32.825Z [TRACE] readResourceInstanceState: no state present for aws_alb.alb01
2021-12-02T11:07:32.825Z [TRACE] reducePlan: aws_alb.alb01 change simplified from DeleteThenCreate to Create for apply node
2021-12-02T11:07:32.825Z [INFO]  Starting apply for aws_alb.alb01
2021-12-02T11:07:32.827Z [DEBUG] aws_alb.alb01: applying the planned Create change
2021-12-02T11:07:32.828Z [TRACE] GRPCProvider: ApplyResourceChange


the workaround is rolling back to v3.66 
Joe Negron ~ NYC

**Comment by anGie44** [comment]
Hi @wwwizards, do you mind opening a new github issue with more configuration details regarding the errors you are seeing related to `waf.fail_open.enabled`? thank you!

**Comment by wwwizards** [comment]
hi @anGie44 - as requested, I created a [new issue #22037](https://github.com/hashicorp/terraform-provider-aws/issues/22037) for this. Thanks.

**Pull request merged**